### PR TITLE
API: Mimic PureComponent behavior for Consumer children

### PR DESCRIPTION
**Pull request by benoitdion**
This fixes a bug in react-native-server where `SET_CURRENT_STORY` was getting spammed over and over for the same story.

### CODE REVIEW

File: app/react-native-server/src/client/manager/provider.js [hunk 1]
```diff
@@ -76,8 +62,7 @@ export default class ReactProvider extends Provider {
   handleAPI(api) {
     addons.loadAddons(api);
 
-    api.onStory((kind, story) => {
-      this.selection = { kind, story };
+    api.onStory(() => {
```
Comment (ndelangen) [inline app/react-native-server/src/client/manager/provider.js]: `api.onStory` should at some point be changed to `api.on`

I understand onStory is kinda convenient, but I don't want to add a method for every event.

File: lib/api/src/index.tsx [hunk 2]
```diff
@@ -233,8 +238,17 @@ class ManagerConsumer extends Component<ConsumerProps<SubState, Combo>> {
...
           const data = this.dataMemory ? this.dataMemory(d) : d;
-
-          return children(data);
+          if (
+            this.props.pure &&
+            this.prevChildren &&
+            this.prevData &&
+            shallowEqualObjects(data, this.prevData)
+          ) {
+            return this.prevChildren;
```
[the remark this answers is not in the document]
Comment (benoitdion) [reply lib/api/src/index.tsx]: > later I realized this mean ALL descendant components would have to be pure for this to work as expected.

Why is that the case? Or, what's preventing descendants to not be pure?

[the remark this answers is not in the document]
Comment (ndelangen) [reply lib/api/src/index.tsx]: @benoitdion if you return prevChildren from render, any updates to the new children won't be part of the newly created XML tree.

[the remark this answers is not in the document]
Comment (benoitdion) [reply lib/api/src/index.tsx]: I'm still not sure I fully understand the problem. I added 7b2539a7f4dd74e6f383f561302636878a3165fa to show an example of a descendant updating when the provider data doesn't change. 

When the provider data changes, `shallowEqualObjects` should return false and  the`children` function will get called again.

**Comment by ndelangen** [comment]
I'll fix the merge conflict & linting today and merge

**Comment by benoitdion** [comment]
I'll take early next week if that's ok. There are a couple of things I wanted to clean up before merging.

**Pull request merged**